src/main.rs

```diff
@@ -178,44 +178,34 @@ enum TableError {
     Pager(PagerError),
 }
 
-impl Table {
-    fn new<P>(filename: P) -> Result<Self, TableError>
-    where
-        P: AsRef<Path>,
-    {
-        Pager::new(filename)
-            .map_err(TableError::Pager)
-            .map(|pager| {
-                let num_rows: u32 = (pager.file_length / ROW_SIZE as u64) as u32;
-                Table { num_rows, pager }
-            })
-    }
+struct VirtualMachine<'a> {
+    table: &'a mut Table,
+}
 
+impl VirtualMachine<'_> {
     fn execute_statement<'a>(
         &'a mut self,
         statement: Statement<'a>,
     ) -> Result<ReplResult, ExecuteError> {
         match statement {
             Statement::Insert { row } => {
-                if self.num_rows == TABLE_MAX_ROWS as u32 {
+                if self.table.num_rows == TABLE_MAX_ROWS as u32 {
```

Comment:
Technically the goal of the writer was to keep table from knowing anything about how data is stored. In my original implementation, I believe this was already true, but I still wanted to implement the same design (as closely as possible except where I felt subjectively interested in some other design).